HHS/Head-Start-TTADP
# [TTAHUB-2439] Grant replacement: models and migrations

## Description of change

Support for 1->n and n->1 grant replacements by introducing three new tables:

- `GrantReplacements`
- `GrantReplacementTypes`
- `GrantRelationshipToActive`

These are fairly straight-forward constructs, so looking at the models should tell you everything you need to know about them and how they are intended to be used.

The bulk of the changes are in

- `processFiles` in `src/lib/updateGrantsRecipients.js`
- `mergeGoals` in `src/goalServices/goals.js`
- various handlers/services that query the `Grants` table or referenced `oldGrantId`

The changes to the `Grant` `filtersToScopes` are not strictly necessary any longer since we decided to keep the `inactivationDate` and `inactivationReason` (this meant no join was necessary), but I left it there because I think it could be useful in the future. For example, if we wanted to create a new filter that returns all grants that a particular grant has replaced, we could do so without writing any raw SQL in the filter by leveraging the `return { includes: [] }` pattern.

### inactivationDate and inactivationReason

It seems we need to keep these because they *are not the same* as `replacementDate` and `replacementType`. For example:

This is an excerpt from `grant_award.xml`:

```xml
  <grant_award_id>7842</grant_award_id>
  ...
  <grant_status>Inactive</grant_status>
  <inactivation_date>2022-07-31</inactivation_date>
  <inactivation_reason>Replaced</inactivation_reason>
```

And from `grant_award_replacement.xml` for the same grant (7842):

```xml
<grant_award_replacement>
    <replaced_grant_award_id>7842</replaced_grant_award_id>
    <replacement_grant_award_id>8110</replacement_grant_award_id>
    <replacement_date>10-01-2021</replacement_date>
    <grant_replacement_type>DRS Non-Competitive Continuation</grant_replacement_type>
</grant_award_replacement>
```

Note the inactivation date is not the same as the replacement date, and the inactivation reason is not the same as the replacement type.

`inactivation_reason` -> `Grant.inactivationReason`
`grant_replacement_type` -> unique `GrantReplacementTypes.name`

`inactivation_date` -> `Grant.inactivationDate`
`replacement_date` -> `GrantReplacement.replacementDate`

New tests were added for the `processFiles` method to validate this.

## How to test

- Walk through the logic in `src/lib/updateGrantsRecipients.js` for the `processFiles` function. This still populates columns on `Grant` like `inactivationDate` and `inactivationReason`, but it now also creates `GrantReplacement` and `GrantReplacementType` entities. After an import, check your database to make sure these rows are created.
- There's no changes to the UI, so the only tests we can do are make sure everything still works, which I think the e2e tests do a pretty good job of.
- Validate the changes in `src/goalServices/goals.js` for the `mergeGoals` function. It should grab all old goals from all grants being replaced, not just the one previously referenced by the now-removed `oldGrantId` column.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-2439
* https://jira.acf.gov/browse/TTAHUB-2440
* https://jira.acf.gov/browse/TTAHUB-2441
* https://jira.acf.gov/browse/TTAHUB-2442
* https://jira.acf.gov/browse/TTAHUB-2443
* https://jira.acf.gov/browse/TTAHUB-2444
* https://jira.acf.gov/browse/TTAHUB-2445


## Checklists

### Every PR



### Before merge to main


### Production Deploy


### After merge/deploy
